datafusion/optimizer/src/type_coercion.rs

```diff
@@ -103,7 +103,9 @@ fn optimize_internal(
             // ensure aggregate names don't change:
             // https://github.com/apache/arrow-datafusion/issues/3555
             if matches!(expr, Expr::AggregateFunction { .. }) {
-                if let Some((alias, name)) = original_name.zip(expr.name().ok()) {
+                if let Some((alias, name)) =
+                    original_name.zip(Some(expr.canonical_name().replace("#", "")))
```

Comment:
`expr.name()` and `expr.canonical_name()` differ in the use of `#` to prefix a column name, hence the need for a hack here to strip this character.

Is it intentional that we have this difference between the two methods?

Reply:
We have found the need for this hack in Dask SQL as well.

Reply:
I do not know what the difference is -- maybe @Dandandan knows

Reply:
I think we should make the behavior consistent between the two ways of building a name. This seems like an oversight rather than an intentional difference.

Reply:
Filed https://github.com/apache/arrow-datafusion/issues/3711 for this

Reply:
Once https://github.com/apache/arrow-datafusion/pull/3712 is merged, the `replace("#", "")` hack can be removed.